flink-ml-core/src/main/java/org/apache/flink/ml/util/QuantileSummary.java

```diff
@@ -0,0 +1,359 @@
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Helper class to compute an approximate quantile summary. This implementation is based on the
+ * algorithm proposed in the paper: "Space-efficient Online Computation of Quantile Summaries" by
+ * Greenwald, Michael and Khanna, Sanjeev. (https://doi.org/10.1145/375663.375670)
+ */
+public class QuantileSummary {
+
+    private static final int BUFFER_SIZE = 10000;
+    private static final int COMPRESS_THRESHOLD = 10000;
+
+    private double relativeError;
+    private long count;
+    private List<StatsTuple> sampled;
+    private List<Double> headBuffer = new ArrayList<>(BUFFER_SIZE);
+
+    /**
+     * * QuantileSummaries Constructor.
+     *
+     * @param relativeError The target relative error.
+     */
+    public QuantileSummary(double relativeError) {
+        this(relativeError, Collections.EMPTY_LIST, 0);
+    }
+
+    /**
+     * QuantileSummaries Constructor.
+     *
+     * @param relativeError The target relative error.
+     * @param sampled A buffer of quantile statistics. See the G-K article for more details.
+     * @param count The count of all the elements inserted in the sampled buffer.
+     */
+    public QuantileSummary(double relativeError, List<StatsTuple> sampled, long count) {
+        if (relativeError <= 0 || relativeError >= 1) {
+            throw new RuntimeException("An appropriate relative error must lay between 0 and 1.");
+        }
+        this.relativeError = relativeError;
+        this.sampled = sampled;
+        this.count = count;
+    }
+
+    /**
+     * Insert a new observation to the summary.
+     *
+     * @param item The new observation to insert into the summary.
+     * @return A summary with the given observation inserted into the summary.
+     */
+    public QuantileSummary insert(Double item) {
+        headBuffer.add(item);
+        if (headBuffer.size() >= BUFFER_SIZE) {
+            QuantileSummary result = insertHeadBuffer();
+            if (result.sampled.size() >= COMPRESS_THRESHOLD) {
+                return result.compress();
+            } else {
+                return result;
+            }
+        } else {
+            return this;
+        }
+    }
+
+    /**
+     * This implements the COMPRESS function of the GK algorithm.
+     *
+     * @return A new summary that compresses the summary statistics and the head buffer.
+     */
+    public QuantileSummary compress() {
+        QuantileSummary inserted = insertHeadBuffer();
+        Preconditions.checkState(inserted.headBuffer.isEmpty());
+        Preconditions.checkState(inserted.count == count + headBuffer.size());
```

Comment:
Is `Preconditions.checkState(inserted.count == count)` enough since headBuffer is empty?

Reply:
It is `inserted.headBuffer` which is empty, not `headBuffer`, so I think the check statement is right.